### ISSUE TITLE
IRCClient.msg loses data when splitting long messages

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5176]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)| @jonathanj reported|
|-|-|
|Trac ID|trac#5176|
|Type|enhancement|
|Created|2011-07-04 17:01:54Z|
|Branch|https://github.com/twisted/twisted/tree/fix-ircclient-msg-split-5176-2|

The length calculations used for splitting long messages don't take into account that when a message like PRIVMSG #foo <long_message> is relayed to other clients by the server it looks something like :nickname!user@host PRIVMSG #foo :<long_message> which is truncated at 510 characters. The current behaviour means that <long_message> loses an unspecified number of characters around the message limit mark for each line.

I think we have enough information to guess a safe maximum:

 * Maximum nickname length can be determined from the ISUPPORT info (using our own current nickname may not always be correct.)
 * Maximum username length seems to be something like 10 characters, I can't find any information on this limit.
 * According to RFC 2812 the maximum hostname is 63 characters (our hostname can be changed on the server side at any point.)

<details><summary>Searchable metadata</summary>

```
trac-id__5176 5176
type__enhancement enhancement
reporter__jonathanj jonathanj
priority__normal normal
milestone__ 
branch__branches_fix_ircclient_msg_split_5176_2 branches/fix-ircclient-msg-split-5176-2
branch_author__jonathanj jonathanj
status__closed closed
resolution__fixed fixed
component__words words
keywords__irc irc
time__1309798914000000 1309798914000000
changetime__1311865507000000 1311865507000000
version__None None
owner__jonathanj jonathanj
cc__ralphm cc__jerith
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_2"></a>|@jonathanj commented|
|-|-|

(In [32230]) Branching to 'fix-ircclient-msg-split-5176'

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_3"></a>|@jonathanj commented|
|-|-|

(In [32231]) Docstrings and cosmetics for twisted.words.test.test_irc.ClientMsgTests, refs [#5176](https://github.com/twisted/twisted/issues/5176).

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_4"></a>|@jonathanj commented|
|-|-|

(In [32232]) Fix what appears to be a copy-paste error, refs [#5176](https://github.com/twisted/twisted/issues/5176).

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_5"></a>|@jonathanj commented|
|-|-|

(In [32234]) Write better tests for IRCClient.msg splitting behaviour, refs [#5176](https://github.com/twisted/twisted/issues/5176).

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_6"></a>|@jonathanj commented|
|-|-|

(In [32235]) Better maximum command length calculation to make failing tests pass, refs [#5176](https://github.com/twisted/twisted/issues/5176).

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_8"></a>|@jonathanj commented|
|-|-|

Buildbot results: <http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/fix-ircclient-msg-split-5176>

**Comment by twisted-trac** [comment]
|[<img alt="jerith's avatar" src="https://avatars.githubusercontent.com/u/43731?s=50" width="50" height="50">](https://github.com/jerith)<a name="note_9"></a>|@jerith set owner to @jonathanj|
|-|-|

Disclaimer: This is my first actual review, so yell if I'm doing it wrong.

(There's a merge conflict with current trunk r32338 ([#4990](https://github.com/twisted/twisted/issues/4990) -- assertEquals->assertEqual), so I used the version from the branch which includes the same changes.)

Why do we use the generic maximum value for the command prefix (which is our own nick!user@host) instead of calculating the actual length? Do we not have reliable access to that?

Apart from those, I don't see any problems with merging the branch.

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_10"></a>|@jonathanj commented|
|-|-|

(In [32375]) Branching to 'fix-ircclient-msg-split-5176-2'

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_11"></a>|@jonathanj commented|
|-|-|

Replying to [jerith](#note_9):
> (There's a merge conflict with current trunk r32338 ([#4990](https://github.com/twisted/twisted/issues/4990) -- assertEquals->assertEqual), so I used the version from the branch which includes the same changes.)

Fixed.

> Why do we use the generic maximum value for the command prefix (which is our own nick!user@host) instead of calculating the actual length? Do we not have reliable access to that?

Yes, it is not reliable enough. I've expanded the docstring for _safeMaximumLength to better explain this.

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)|@jonathanj removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jerith's avatar" src="https://avatars.githubusercontent.com/u/43731?s=50" width="50" height="50">](https://github.com/jerith)<a name="note_13"></a>|@jerith set owner to @jonathanj|
|-|-|

Looks good for merging.

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_14"></a>|@jonathanj set status to `closed`|
|-|-|

(In [32380]) Merge fix-ircclient-msg-split-5176-2.

Author: jonathanj
Reviewer: jerith
Fixes: [#5176](https://github.com/twisted/twisted/issues/5176)

Be more conservative when determining message length in IRCClient.msg in order to reduce the chance of relayed IRC messages being truncated.
